src/main/java/fi/aalto/cs/apluscourses/presentation/CourseSelectionViewModel.java

```diff
@@ -0,0 +1,37 @@
+package fi.aalto.cs.apluscourses.presentation;
+
+import fi.aalto.cs.apluscourses.utils.observable.ObservableProperty;
+import fi.aalto.cs.apluscourses.utils.observable.ObservableReadWriteProperty;
+import fi.aalto.cs.apluscourses.utils.observable.ValidationError;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+
+public class CourseSelectionViewModel {
+
+  private List<CourseItemViewModel> courses;
+
+  public final ObservableProperty<String> selectedCourseUrl
+      = new ObservableReadWriteProperty<>("", CourseSelectionViewModel::validateCourseUrl);
+
+  public CourseSelectionViewModel(List<CourseItemViewModel> courses) {
+    this.courses = courses;
+  }
+
+  public CourseItemViewModel[] getCourses() {
+    return courses.toArray(new CourseItemViewModel[0]);
+  }
+
+  private static ValidationError validateCourseUrl(String courseUrl) {
+    if (courseUrl.trim().isEmpty()) {
+      return () -> "Select a course";
```

Comment:
A couple of nested classes would fit so well in this class but okay...

Reply:
:joy: I was so close to changing the existing nested classes to lambdas as well...